### ISSUE TITLE
Metrics for Grafana/Prometheus

### ISSUE THREAD

**Issue by MrSipping** [nginx/unit#837]
Hi, how can i retrieve some metrics to pass later on grafana/prometheus. I can't find any information on the site

**Comment by lcrilly** [comment]
Did you look at https://unit.nginx.org/usagestats/ ?

Here's a simple Python script to extract the metrics in Prometheus format. You can configure this as an `"application"` if the control socket has a TCP port.

```python
import os
import http.client
import json

def application(environ, start_response):
    connection = http.client.HTTPConnection(os.getenv('ADDR'))
    connection.request("GET", "/status")
    response = connection.getresponse()

    statj = json.loads(response.read())
    connection.close()

    metrics = []
    metrics.append("unit_connections_accepted_total {}".format(statj["connections"]["accepted"]))
    metrics.append("unit_connections_active {}".format(statj["connections"]["active"]))
    metrics.append("unit_connections_idle {}".format(statj["connections"]["idle"]))
    metrics.append("unit_connections_closed_total {}".format(statj["connections"]["closed"]))
    metrics.append("unit_requests_total {}".format(statj["requests"]["total"]))

    for application in statj["applications"].keys():
        metrics.append("unit_application_" + application + "_processes_running {}".format(statj["applications"][application]["processes"]["running"]))
        metrics.append("unit_application_" + application + "_processes_starting {}".format(statj["applications"][application]["processes"]["starting"]))
        metrics.append("unit_application_" + application + "_processes_idle {}".format(statj["applications"][application]["processes"]["idle"]))
        metrics.append("unit_application_" + application + "_requests_active {}".format(statj["applications"][application]["requests"]["active"]))

    start_response('200 OK', [("Content-Type", "text/plain")])
    yield str.encode("\n".join(metrics) + "\n")
```

**Comment by tkumark** [comment]
I am new to this. Can you suggest this how to implement this when using unit as container. I am evaluating unit with nodejs in a container. How do I run above scirpt for a container solution.

**Comment by tippexs** [comment]
Hi @tkumark for what ever reason this issue slipped of my radar! Sorry for that. The code shared by @lcrilly can be configured as a python application. More information about this. http://unit.nginx.org/configuration/#python

This `application` can be attached to a `listener` for example on port `9090.

There is furthermore no problem having multiple applications on Unit in a single container.

Let me know if that helps and in case you have any further questions do not hesitate to leave another comment in this thread.

**Comment by meezaan** [comment]
> Hi @tkumark for what ever reason this issue slipped of my radar! Sorry for that. The code shared by @lcrilly can be configured as a python application. More information about this. http://unit.nginx.org/configuration/#python
> 
> This `application` can be attached to a `listener` for example on port `9090.
> 
> There is furthermore no problem having multiple applications on Unit in a single container.
> 
> Let me know if that helps and in case you have any further questions do not hesitate to leave another comment in this thread.

What user would this application need to run as. When running in a container I can ssh into the container and ```curl --unix-socket /var/run/control.unit.sock http://localhost/status``` and get the metrics (but obviously that's the `root` user. If I run the same command as an application user or using a Python or PHP I'm not running as root and so the request fails (with a curl error 7).

**Comment by meezaan** [comment]
OK, I may have a solution. Watch this space!

**Comment by meezaan** [comment]
OK, I stand corrected. I have a running sidecar, but that means I have to add another app in my main app folder to expose the stats, and ideally that's not the case. @lcrilly @tippexs would it be possible to have a route in the unit conf file that exposes the states, but only allowed on 127.0.0.1, for instance? Without that being exposed, at least I'm not able to see how I could deploy a complete independent prometheus exporter as a sidecar. Thank you!

**Comment by lcrilly** [comment]
Wrote up a more complete solution for exposing a Prometheus endpoint with the default Unix socket control port.

https://gist.github.com/lcrilly/ad516de8378dd8ae5303a942e67d55b5

**Comment by meezaan** [comment]
I'll test out it out next week. Thanks @lcrilly.

**Comment by lcrilly** [comment]
@meezaan did this work for you?

**Comment by meezaan** [comment]
> @meezaan did this work for you?

@lcrilly Sorry I have not got around to it, but it's now the highest item on my list (so I'll tackle it either this weekend or early next week). So expect something here early next week.

**Comment by chopanovv** [comment]
> meezaan

is it works? for everyone?

**Comment by meezaan** [comment]
> > meezaan
> 
> is it works? for everyone?

No. Not yet. Still does not have what I believe is needed to run an independent sidecar.
